python/ray/data/tests/test_numpy_support.py

```diff
@@ -125,39 +149,64 @@ def test_ragged_array_like(ray_start_regular_shared):
     )
 
 
-def test_scalar_nested_arrays(ray_start_regular_shared):
+def test_scalar_nested_arrays(ray_start_regular_shared, restore_data_context):
+    # Disable (automatic) fallback to `ArrowPythonObjectType` extension type
+    DataContext.get_current().enable_fallback_to_arrow_object_ext_type = False
+
     data = [[[1]], [[2]]]
     output = do_map_batches(data)
-    assert_structure_equals(output, create_ragged_ndarray(data))
 
+    assert_structure_equals(
+        output,
+        create_ragged_ndarray(
+            [np.array([1], dtype=np.object_), np.array([2], dtype=np.object_)]
+        ),
+    )
+
+
+def test_scalar_lists_not_converted(ray_start_regular_shared, restore_data_context):
+    # Disable (automatic) fallback to `ArrowPythonObjectType` extension type
+    DataContext.get_current().enable_fallback_to_arrow_object_ext_type = False
 
-def test_scalar_lists_not_converted(ray_start_regular_shared):
     data = [[1, 2], [1, 2]]
     output = do_map_batches(data)
-    assert_structure_equals(output, create_ragged_ndarray([[1, 2], [1, 2]]))
+    assert_structure_equals(
+        output, create_ragged_ndarray([np.array([1, 2]), np.array([1, 2])])
+    )
 
     data = [[1, 2, 3], [1, 2]]
     output = do_map_batches(data)
-    assert_structure_equals(output, create_ragged_ndarray([[1, 2, 3], [1, 2]]))
```

Comment:
This would previously be a result of 

 - We'd try to serialize as tensor, this would fail
 - We fallback to `ArrowPythonObjectArray` (inside `numpy_to_block` method) that would pickle provided lists instead of using Arrow native encoder (resulting into subpar performance)

Now this would be serialized to Arrow's native array type using `_table_from_pydict` unifying this flow with the way BlockBuilders convert batches.

Reply:
In other words this would previously not work as expected resulting in silent failure like following:

```
2024-11-06 18:05:42,050 DEBUG arrow_block.py:251 -- Failed to interpret output as multi-dimensional arrays. It will be pickled.
```